junifer/storage/hdf5.py

```diff
@@ -0,0 +1,954 @@
+        )
+
+    def store_timeseries(
+        self,
+        meta_md5: str,
+        element: Dict[str, str],
+        data: np.ndarray,
+        col_names: Optional[Iterable[str]] = None,
+    ) -> None:
+        """Store timeseries.
+
+        Parameters
+        ----------
+        meta_md5 : str
+            The metadata MD5 hash.
+        element : dict
+            The element as dictionary.
+        data : numpy.ndarray
+            The timeseries data to store.
+        col_names : list or tuple of str, optional
+            The column labels (default None).
+
+        """
+        self._store_data(
+            kind="timeseries",
+            meta_md5=meta_md5,
+            element=[element],  # convert to list
+            data=data[:, :, np.newaxis],  # convert to 3D
+            column_headers=col_names,
+            row_header_column_name="timepoint",
+        )
+
+    def collect(self) -> None:
+        """Implement data collection.
+
+        This method globs the element files and runs a loop
+        over them while reading metadata and then runs a loop
+        over all the stored features in the metadata, storing
+        the metadata and the feature data right after reading.
+
+        Raises
+        ------
+        NotImplementedError
+            If ``single_output`` is True.
+
+        """
+        if self.single_output is True:
+            raise_error(
+                msg="collect() is not implemented for single output.",
+                klass=NotImplementedError,
+            )
+
+        # Glob files
+        globbed_files = self.uri.parent.glob(  # type: ignore
+            f"*{self.uri.name}"  # type: ignore
+        )
+
+        # Create new storage instance
+        out_storage = HDF5FeatureStorage(uri=self.uri, overwrite="update")
+
+        # Run loop to collect metadata
+        logger.info(
+            "Collecting metadata from "
+            f"{self.uri.parent}/*{self.uri.name}"  # type: ignore
+        )
+        # Collect element files per feature MD5
+        elements_per_feature_md5 = defaultdict(list)
+        for file_ in tqdm(globbed_files, desc="file-metadata"):
+            logger.debug(f"Reading HDF5 file: {file_} ...")
+            # Create new storage instance to load metadata
+            in_storage = HDF5FeatureStorage(uri=file_)
+            # Load metadata from new instance
+            in_metadata = in_storage._read_metadata()
+
+            logger.info(f"Updating HDF5 metadata with metadata from: {file_}")
+            # Load metadata; empty dictionary if first entry;
+            # can be replaced with store_metadata() if run on a loop
+            # for the metadata entries from in_storage
+            try:
+                out_metadata = out_storage._read_metadata()
+            except IOError:
+                out_metadata = {}
+            # Update metadata
+            out_metadata.update(in_metadata)
+            # Save metadata
+            out_storage._write_processed_data(
```

Comment:
Why writing here? For each on of the 40k files, for each feature, you are triggering a write.

This should be done on the second step of the collect.

Reply:
Looks like an old review, resolving.